File: app/src/main/java/com/manoj/baseproject/utils/helper/SingleRequestEvent.java
package com.manoj.baseproject.utils.helper;


import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;

import com.manoj.baseproject.utils.helper.Resource;
import com.manoj.baseproject.utils.helper.SingleLiveEvent;


public class SingleRequestEvent<T> extends SingleLiveEvent<Resource<T>> {
    public interface RequestObserver<T> {
        void onRequestReceived(@NonNull Resource<T> resource);
    }

    public void observe(LifecycleOwner owner, final RequestObserver<T> observer) {
        super.observe(owner, resource -> {
            if (resource == null) {
                return;
            }

            observer.onRequestReceived(resource);
        });
    }

}

File: app/src/main/java/com/manoj/baseproject/core/network/helper/SingleRequestEvent.kt
package com.manoj.baseproject.core.network.helper

import androidx.lifecycle.LifecycleOwner

class SingleRequestEvent<T> : SingleLiveEvent<Result<T?>>() {

    interface RequestObserver<T> {
        fun onRequestReceived(result: Result<T?>)
    }

    fun observe(owner: LifecycleOwner, observer: RequestObserver<T>) {
        super.observe(owner) { result ->
            result.let {
                observer.onRequestReceived(it)
            }
        }
    }
}
